cyblog.md

```diff
@@ -0,0 +1,75 @@
+    records: [certCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordValue)],
+    ttl: 60,
+    type: certCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordType),
+    zoneId: zone.apply(zone => zone.id),
+});
+```
+
+In particular, creating the `aws.route53.Record` involve a fair amount of complexity with those arrow functions.  i.e.:
+
+```ts
+    records: [certCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordValue)],
+    ttl: 60,
+    type: certCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordType),
+    zoneId: zone.apply(zone => zone.id),
+```
+
+Yikes!  This is so verbose, it doesn't even fit on the width of the page!
+
+The idea of the `.apply` function is similar to `Promise.then`.  It allows one to pass a piece of code that will be applied to the underlying value, and will return an `Output` that then points to the transformed underlying value.  Importantly, the new `Output` will still track dependency information properly.  In the above example the `aws.route53.Record` will know that it both depends on `certCertificate` and on `zone`, even though neither of those resources (or their direct Outputs) are passed directly do the constructor.  
+
+Unfortunately, while `.apply` gives a lot of power and flexibility, it is also somewhat verbose and clunky for describing such a simple concept.  Fortunately, we found a way to improve the situation greatly.  This realization came about from great work done in our Python package.  First, before diving into the low level details, let's first see what the above code would now look like in 0.17.0:
+
+```ts
+// certCertificate and zone created as above
+
+const certValidation = new aws.route53.Record("cert_validation", {
+    records: [certCertificate.domainValidationOptions[0].resourceRecordValue],
+    ttl: 60,
+    type: certCertificate.domainValidationOptions[0].resourceRecordType,
+    zoneId: zone.id,
+});
+```
+
+That's a lot nicer than before!  The following improvements happened:
+1. there's no more lambda!
+2. there's no need for a repetive lambda parameter (which might conflict with some other name in scope).
+3. the code *is* just pure simple *idiomatic* JavaScript/TypeScript that clearly conveys its intent.
+
+Importantly, no information has been lost here.  The exact same dependency information flows along here like it did before.  And, thanks to TypeScript's flexible type system, the above is totally typesafe and will let you know the right types of things and will still error if you happen to make mistakes.
+
+So, how was this done?  Well, the core part of the change is thanks to a little-known feature of JavaScript: (Proxies)[https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy].
```

Comment:
I think something funny is happening with the Markdown here, GitHub doesn't seem to render this link correctly?

Reply:
I think the `(`s and `)s` got swapped with the `[`s and `]`s.